### DISCUSSION TITLE
데이터 수 조회 기능 개선

### DISCUSSION THREAD

**Discussion by woo-chang** [Busan-Dinosaur/Foodbowl-Backend#122] [개선해볼까요?]
현재 유저 정보를 확인하는 상황이라면 `팔로워 수`도 같이 조회되도록 되어있습니다. 테이블 구조는 아래와 같습니다.

<img width="961" alt="스크린샷 2023-08-12 오후 7 35 52" src="https://github.com/Busan-Dinosaur/Foodbowl-Backend/assets/78212016/3e94b953-2b93-45ba-a52f-7e72f76c83de">

멤버와 팔로우는 `1:N`의 관계로 팔로워 수를 알기 위해서는 조인이 필요합니다. 회원 정보를 조회할 뿐인데 매번 조인을 통해 팔로워 수를 조회하는 것은 성능적인 문제가 발생할 수 있습니다. 이를 해결하는 방법에는 캐시, 조회용 NoSQL, 반정규화 등의 방법이 존재합니다.

캐시나 조회용 NoSQL을 두는 것도 좋지만, 생각보다 비용이 많이 발생할 것 같기에 반정규화를 통해 이 문제를 해결해나가고자 합니다.

멤버 필드에 팔로워 수 필드를 추가하여 팔로워 관련 기능이 수행될 때마다 팔로워 수 필드를 업데이트시키는 방법을 사용할 수 있습니다. 팔로워 수와 팔로워 데이터 정합성을 꼼꼼히 확인해주어야 한다는 문제도 있지만, 코드 레벨을 통해 동시성과 정합성을 해결할 수 있고, 팔로워 수가 완전히 정확해야 할 만큼 중요한 데이터도 아니기도 합니다. 필요하다면 일정 시간마다 정합성을 맞춰주는 작업을 추가해도 좋습니다.

결론적으로 멤버에 팔로워 수 필드를 추가하는 것에 대해 얘기해보고 싶습니다 :)

**Comment by Kim0914** [comment]
회원 정보를 조회할 뿐인데 매번 조인을 통해 팔로워 수를 조회하는 부분에 대해서 매우 공감합니다 ㅎㅎ

또한 현재 저희 서비스가 네트워킹(팔로우 / 팔로워) 기반 서비스이기 때문에 가장 조회가 많이 발생할 구간이라고 생각도 들어요 !

말씀해주신 것처럼 캐시나 조회용 DB를 두는 것 보다, 데이터의 중복을 허용하더라도 반정규화를 진행하는게 더 적절할 것 같구요.

구글링과 몇 개의 포스팅들을 찾아보았는데, 반정규화는 테이블의 정합성을 깰 수도 있는 행위이기 때문에 신중하게 도입해야 한다고 합니다.

예를 들어, 반정규화를 도입하기 전에 `조인 테이블 수`, `해당 쿼리를 처리하는데 소요되는 시간`, `성능이 얼마나 떨어지는지` 등의 지표를 조사하고,
다른 방법들(**인덱스 조절**, **뷰 테이블 도입** 등)을 고려한 이후에 반정규화를 진행한다고 합니다 !

결론적으로, 반정규화를 도입한다고 하면 **찬성**이지만 도입한 근거와 통계 자료들을 확실히 남겨두어야 할 것 같아요 !

추가적으로 현재 기능 개발 기간이 많이 남지 않은 것 같은데, 이 부분도 함께 고려해보면 좋겠네요 😁

**Reply by woo-chang** [reply]
말씀해주신대로 반정규화를 도입함으로 얻는 **단점**(데이터 정합성 문제 등)에 대해서도 고려 후에 개선이 필요한 부분인 것 같아요!
당장은 현재 테이블 구조로도 충분히 문제를 풀어낼 수 있는 상황이기에 모든 기능이 완료된 후 같이 고민해보면 좋을 것 같아요 :)

답변 감사합니다 🙇🏻‍♂️
